File: Unreal/Source/ToS_Network/Private/Network/SecureSession.cpp
#include "Network/SecureSession.h"
#include "Utils/LZ4.h"
#include "Utils/FileLogger.h"
#include "HAL/UnrealMemory.h"

static FString BytesToHexString(const TArray<uint8>& Bytes)
{
    FString Result;
    Result.Reserve(Bytes.Num() * 2);

    for (uint8 Byte : Bytes)
    {
        Result += FString::Printf(TEXT("%02X"), Byte);
    }

    return Result;
}

void FPacketHeader::Serialize(uint8* Buffer) const
{
    FMemory::Memcpy(Buffer, &ConnectionId, 4);
    Buffer[4] = static_cast<uint8>(Channel);
    Buffer[5] = static_cast<uint8>(Flags);
    FMemory::Memcpy(Buffer + 6, &Sequence, 8);
}

FPacketHeader FPacketHeader::Deserialize(const uint8* Buffer)
{
    FPacketHeader Header;
    FMemory::Memcpy(&Header.ConnectionId, Buffer, 4);
    Header.Channel = static_cast<EPacketChannel>(Buffer[4]);
    Header.Flags = static_cast<EPacketHeaderFlags>(Buffer[5]);
    FMemory::Memcpy(&Header.Sequence, Buffer + 6, 8);
    return Header;
}

TArray<uint8> FPacketHeader::GetAAD() const
{
    TArray<uint8> AAD;
    AAD.SetNumUninitialized(Size);
    Serialize(AAD.GetData());
    return AAD;
}

bool FSecureSession::InitializeAsClient(const TArray<uint8>& ClientPrivateKey, const TArray<uint8>& ServerPublicKey,
                                       const TArray<uint8>& Salt, uint32 InConnectionId)
{
    if (ClientPrivateKey.Num() != 32 || ServerPublicKey.Num() != 32 || Salt.Num() != 16)
        return false;

    uint8 SharedSecret[32];
    if (crypto_scalarmult_curve25519(SharedSecret, ClientPrivateKey.GetData(), ServerPublicKey.GetData()) != 0)
        return false;

    uint8 PRK[32];
    crypto_auth_hmacsha256_state state;
    crypto_auth_hmacsha256_init(&state, Salt.GetData(), 16);
    crypto_auth_hmacsha256_update(&state, SharedSecret, 32);
    crypto_auth_hmacsha256_final(&state, PRK);

    const char* Info = "ToS-UE5 v1";
    uint8 OKM[64];
    uint8 T[32];

    crypto_auth_hmacsha256_init(&state, PRK, 32);
    crypto_auth_hmacsha256_update(&state, (const uint8*)Info, strlen(Info));
    crypto_auth_hmacsha256_update(&state, (const uint8*)"\x01", 1);
    crypto_auth_hmacsha256_final(&state, T);
    FMemory::Memcpy(OKM, T, 32);

    crypto_auth_hmacsha256_init(&state, PRK, 32);
    crypto_auth_hmacsha256_update(&state, T, 32);
    crypto_auth_hmacsha256_update(&state, (const uint8*)Info, strlen(Info));
    crypto_auth_hmacsha256_update(&state, (const uint8*)"\x02", 1);
    crypto_auth_hmacsha256_final(&state, T);
    FMemory::Memcpy(OKM + 32, T, 32);

    FMemory::Memcpy(TxKey, OKM + 32, 32);
    FMemory::Memcpy(RxKey, OKM, 32);

    FMemory::Memcpy(SessionSalt, Salt.GetData(), 16);

    ConnectionId = InConnectionId;
    SeqTx = 0;
    SeqRx = 0;
    ReplayWindow = 0;
    HighestSeqReceived = 0;

    return true;
}

void FSecureSession::GenerateNonce(uint64 Sequence, TArray<uint8>& Nonce) const
{
    Nonce.SetNumUninitialized(12);
    FMemory::Memcpy(Nonce.GetData(), &ConnectionId, 4);
    FMemory::Memcpy(Nonce.GetData() + 4, &Sequence, 8);
}

bool FSecureSession::EncryptPayload(const TArray<uint8>& Plaintext, const TArray<uint8>& AAD, TArray<uint8>& Ciphertext)
{
    TArray<uint8> Nonce;
    GenerateNonce(SeqTx, Nonce);

    // Log crypto data to file (only first packet)
    if (SeqTx == 0)
    {
        ClientFileLogHex(TEXT("[CLIENT] Nonce"), Nonce);
    }



    Ciphertext.SetNumUninitialized(Plaintext.Num() + crypto_aead_chacha20poly1305_ietf_ABYTES);

    unsigned long long CiphertextLen;
    int Result = crypto_aead_chacha20poly1305_ietf_encrypt(
        Ciphertext.GetData(), &CiphertextLen,
        Plaintext.GetData(), Plaintext.Num(),
        AAD.GetData(), AAD.Num(),
        nullptr,
        Nonce.GetData(), TxKey
    );

    if (Result != 0)
    {
        UE_LOG(LogTemp, Error, TEXT("[CRYPTO] Encrypt failed with result: %d"), Result);
        return false;
    }

    Ciphertext.SetNum(CiphertextLen);

    SeqTx++;

    // Log TxKey to file (only once)
    if (SeqTx == 1)
    {
        TArray<uint8> TxKeyArray;
        TxKeyArray.SetNumUninitialized(32);
        FMemory::Memcpy(TxKeyArray.GetData(), TxKey, 32);
        ClientFileLogHex(TEXT("[CLIENT] TxKey"), TxKeyArray);
    }

    return true;
}

bool FSecureSession::EncryptPayloadWithCompression(const TArray<uint8>& Plaintext, const TArray<uint8>& AAD, TArray<uint8>& Result, bool& bWasCompressed)
{
    bWasCompressed = false;

    TArray<uint8> Encrypted;

    if (!EncryptPayload(Plaintext, AAD, Encrypted))
        return false;

    if (Encrypted.Num() > 512)
    {
        TArray<uint8> Compressed;
        Compressed.SetNumUninitialized(Encrypted.Num() + 1024);

        int32 CompressedLength = FLZ4::Compress(Encrypted.GetData(), Encrypted.Num(), Compressed.GetData(), Compressed.Num());

        if (CompressedLength > 0 && CompressedLength < Encrypted.Num())
        {
            bWasCompressed = true;
            Result.SetNumUninitialized(CompressedLength);
            FMemory::Memcpy(Result.GetData(), Compressed.GetData(), CompressedLength);
            return true;
        }
    }

    Result = MoveTemp(Encrypted);
    return true;
}

bool FSecureSession::EncryptPayloadWithSequence(const TArray<uint8>& Plaintext, const TArray<uint8>& AAD, uint64 Sequence, TArray<uint8>& Result, bool& bWasCompressed)
{
    bWasCompressed = false;

    // Use specific sequence instead of internal SeqTx
    TArray<uint8> Nonce;
    GenerateNonce(Sequence, Nonce);

    TArray<uint8> Encrypted;
    Encrypted.SetNumUninitialized(Plaintext.Num() + crypto_aead_chacha20poly1305_ietf_ABYTES);

    unsigned long long CiphertextLen;
    int EncryptResult = crypto_aead_chacha20poly1305_ietf_encrypt(
        Encrypted.GetData(), &CiphertextLen,
        Plaintext.GetData(), Plaintext.Num(),
        AAD.GetData(), AAD.Num(),
        nullptr,
        Nonce.GetData(), TxKey
    );

    if (EncryptResult != 0)
    {
        UE_LOG(LogTemp, Error, TEXT("[CRYPTO] Encrypt failed with result: %d"), EncryptResult);
        return false;
    }

    Encrypted.SetNum(CiphertextLen);

    // Increment SeqTx to keep crypto sequence in sync
    SeqTx++;

    if (Encrypted.Num() > 512)
    {
        TArray<uint8> Compressed;
        Compressed.SetNumUninitialized(Encrypted.Num() + 1024);

        int32 CompressedLength = FLZ4::Compress(Encrypted.GetData(), Encrypted.Num(), Compressed.GetData(), Compressed.Num());

        if (CompressedLength > 0 && CompressedLength < Encrypted.Num())
        {
            bWasCompressed = true;
            Result.SetNumUninitialized(CompressedLength);
            FMemory::Memcpy(Result.GetData(), Compressed.GetData(), CompressedLength);
            return true;
        }
    }

    Result = MoveTemp(Encrypted);
    return true;
}

bool FSecureSession::EncryptPayloadWithSpecificSequence(const TArray<uint8>& Plaintext, const TArray<uint8>& AAD, uint64 Sequence, TArray<uint8>& Ciphertext)
{
    // Use specific sequence instead of internal SeqTx
    TArray<uint8> Nonce;
    GenerateNonce(Sequence, Nonce);

    Ciphertext.SetNumUninitialized(Plaintext.Num() + crypto_aead_chacha20poly1305_ietf_ABYTES);

    unsigned long long CiphertextLen;
    int Result = crypto_aead_chacha20poly1305_ietf_encrypt(
        Ciphertext.GetData(), &CiphertextLen,
        Plaintext.GetData(), Plaintext.Num(),
        AAD.GetData(), AAD.Num(),
        nullptr,
        Nonce.GetData(), TxKey
    );

    if (Result != 0)
    {
        UE_LOG(LogTemp, Error, TEXT("[CRYPTO] Encrypt failed with result: %d"), Result);
        return false;
    }

    Ciphertext.SetNum(CiphertextLen);

    // Increment SeqTx to keep crypto sequence in sync
    SeqTx++;
    return true;
}

bool FSecureSession::DecryptPayload(const TArray<uint8>& Ciphertext, const TArray<uint8>& AAD, uint64 Sequence, TArray<uint8>& Plaintext)
{
    if (!IsSequenceValid(Sequence))
    {
        UE_LOG(LogTemp, Warning, TEXT("[CRYPTO] Decrypt - Replay attack detected! Sequence: %llu"), Sequence);
        return false;
    }

    TArray<uint8> Nonce;
    Nonce.SetNumUninitialized(12);
    FMemory::Memcpy(Nonce.GetData(), &ConnectionId, 4);
    FMemory::Memcpy(Nonce.GetData() + 4, &Sequence, 8);



    Plaintext.SetNumUninitialized(Ciphertext.Num() - crypto_aead_chacha20poly1305_ietf_ABYTES);

    unsigned long long PlaintextLen;
    int Result = crypto_aead_chacha20poly1305_ietf_decrypt(
        Plaintext.GetData(), &PlaintextLen,
        nullptr,
        Ciphertext.GetData(), Ciphertext.Num(),
        AAD.GetData(), AAD.Num(),
        Nonce.GetData(), RxKey
    );

    if (Result != 0)
    {
        UE_LOG(LogTemp, Error, TEXT("[CRYPTO] Decrypt failed with result: %d"), Result);
        return false;
    }

    Plaintext.SetNum(PlaintextLen);

    UpdateReplayWindow(Sequence);
    return true;
}

bool FSecureSession::DecryptPayloadWithDecompression(const TArray<uint8>& Data, const TArray<uint8>& AAD, uint64 Sequence, bool bIsCompressed, TArray<uint8>& Plaintext)
{
    if (bIsCompressed)
    {
        TArray<uint8> Decompressed;
        Decompressed.SetNumUninitialized(Data.Num() * 4);

        int32 DecompressedLength = FLZ4::Decompress(Data.GetData(), Data.Num(), Decompressed.GetData(), Decompressed.Num());
        if (DecompressedLength <= 0)
            return false;

        Decompressed.SetNum(DecompressedLength);
        return DecryptPayload(Decompressed, AAD, Sequence, Plaintext);
    }
    else
    {
        return DecryptPayload(Data, AAD, Sequence, Plaintext);
    }
}

bool FSecureSession::IsSequenceValid(uint64 Sequence) const
{
    if (Sequence > HighestSeqReceived)
        return true;

    if (Sequence + ReplayWindowSize <= HighestSeqReceived)
        return false;

    uint64 Diff = HighestSeqReceived - Sequence;

    if (Diff < ReplayWindowSize)
    {
        uint64 Mask = 1ULL << Diff;
        return (ReplayWindow & Mask) == 0;
    }

    return false;
}

void FSecureSession::UpdateReplayWindow(uint64 Sequence)
{
    if (Sequence > HighestSeqReceived)
    {
        uint64 Shift = Sequence - HighestSeqReceived;

        if (Shift >= ReplayWindowSize)
        {
            ReplayWindow = 1;
        }
        else
        {
            ReplayWindow = (ReplayWindow << Shift) | 1;
        }

        HighestSeqReceived = Sequence;
    }
    else
    {
        uint64 Diff = HighestSeqReceived - Sequence;

        if (Diff < ReplayWindowSize)
        {
            uint64 Mask = 1ULL << Diff;
            ReplayWindow |= Mask;
        }
    }
}

File: Core/Network/Security/SecureSession.cs
using System.Buffers.Binary;
using System.Text;
using Org.BouncyCastle.Crypto.Agreement;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.Crypto.Modes;
using Org.BouncyCastle.Crypto.Macs;

public unsafe struct SecureSession
{
    private static readonly byte[] Info = System.Text.Encoding.ASCII.GetBytes("ToS-UE5 v1");

    public fixed byte TxKey[32];
    public fixed byte RxKey[32];
    public fixed byte SessionSalt[16];
    public ulong SeqTx;
    public ulong SeqRx;
    public uint ConnectionId;

    public readonly string GetTxKeyHex()
    {
        unsafe
        {
            fixed (byte* ptr = TxKey)
            {
                return Convert.ToHexString(new ReadOnlySpan<byte>(ptr, 32));
            }
        }
    }

    public readonly string GetRxKeyHex()
    {
        unsafe
        {
            fixed (byte* ptr = RxKey)
            {
                return Convert.ToHexString(new ReadOnlySpan<byte>(ptr, 32));
            }
        }
    }

    private ulong _replayWindow;
    private ulong _highestSeqReceived;

    public ulong BytesTransmitted;
    public DateTime SessionStartTime;

    private const int ReplayWindowSize = 64;
    private const ulong RekeyBytesThreshold = 1UL << 30; // 1 GB
    private static readonly TimeSpan RekeyTimeThreshold = TimeSpan.FromMinutes(60); // 1 hour

    public static (byte[] serverPublicKey, byte[] salt, SecureSession session) CreateAsServer(ReadOnlySpan<byte> clientPublicKey, uint connectionId)
    {
        var rng = new SecureRandom();
        var serverPriv = new X25519PrivateKeyParameters(rng);
        var serverPub = serverPriv.GeneratePublicKey().GetEncoded();



        var clientPub = new X25519PublicKeyParameters(clientPublicKey);
        var agreement = new X25519Agreement();
        agreement.Init(serverPriv);
        byte[] sharedSecret = new byte[agreement.AgreementSize];
        agreement.CalculateAgreement(clientPub, sharedSecret, 0);

        var salt = new byte[16];
        rng.NextBytes(salt);

        Span<byte> okm = stackalloc byte[64];
        var hkdf = new HkdfBytesGenerator(new Sha256Digest());
        hkdf.Init(new HkdfParameters(sharedSecret, salt, Info));
        hkdf.GenerateBytes(okm);

        SecureSession sess = default;
        okm.Slice(0, 32).CopyTo(new Span<byte>(sess.TxKey, 32));
        okm.Slice(32, 32).CopyTo(new Span<byte>(sess.RxKey, 32));
        salt.CopyTo(new Span<byte>(sess.SessionSalt, 16));
        sess.SeqTx = 0;
        sess.SeqRx = 0;
        sess.ConnectionId = connectionId;
        sess._replayWindow = 0;
        sess._highestSeqReceived = 0;
        sess.BytesTransmitted = 0;
        sess.SessionStartTime = DateTime.UtcNow;

        return (serverPub, salt, sess);
    }

    public static SecureSession CreateAsClient(ReadOnlySpan<byte> clientPrivateKey, ReadOnlySpan<byte> serverPublicKey, ReadOnlySpan<byte> salt, uint connectionId)
    {
        var clientPriv = new X25519PrivateKeyParameters(clientPrivateKey);
        var serverPub = new X25519PublicKeyParameters(serverPublicKey);
        var agreement = new X25519Agreement();
        agreement.Init(clientPriv);
        byte[] sharedSecret = new byte[agreement.AgreementSize];
        agreement.CalculateAgreement(serverPub, sharedSecret, 0);

        Span<byte> okm = stackalloc byte[64];
        var hkdf = new HkdfBytesGenerator(new Sha256Digest());
        hkdf.Init(new HkdfParameters(sharedSecret, salt.ToArray(), Info));
        hkdf.GenerateBytes(okm);

        SecureSession sess = default;
        okm.Slice(32, 32).CopyTo(new Span<byte>(sess.TxKey, 32));   // Client->Server (reversed from server perspective)
        okm.Slice(0, 32).CopyTo(new Span<byte>(sess.RxKey, 32));    // Server->Client
        salt.CopyTo(new Span<byte>(sess.SessionSalt, 16));
        sess.SeqTx = 0;
        sess.SeqRx = 0;
        sess.ConnectionId = connectionId;
        sess._replayWindow = 0;
        sess._highestSeqReceived = 0;
        sess.BytesTransmitted = 0;
        sess.SessionStartTime = DateTime.UtcNow;
        return sess;
    }

    public void GenerateNonce(ulong sequence, Span<byte> nonce)
    {
        if (nonce.Length < 12)
            throw new ArgumentException("Nonce must be at least 12 bytes");

        BinaryPrimitives.WriteUInt32LittleEndian(nonce, ConnectionId);
        BinaryPrimitives.WriteUInt64LittleEndian(nonce.Slice(4), sequence);
    }

    public bool EncryptPayload(ReadOnlySpan<byte> plaintext, ReadOnlySpan<byte> aad, Span<byte> ciphertext, out int ciphertextLength)
    {
        try
        {
            Span<byte> nonce = stackalloc byte[12];
            GenerateNonce(SeqTx, nonce);

            var cipher = new ChaCha20Poly1305();

            unsafe
            {
                fixed (byte* txKeyPtr = TxKey)
                {
                    var keyParam = new KeyParameter(new ReadOnlySpan<byte>(txKeyPtr, 32).ToArray());
                    var parameters = new AeadParameters(keyParam, 128, nonce.ToArray(), aad.ToArray());

                    cipher.Init(true, parameters);
                    ciphertextLength = cipher.ProcessBytes(plaintext.ToArray(), 0, plaintext.Length, ciphertext.ToArray(), 0);
                    ciphertextLength += cipher.DoFinal(ciphertext.Slice(ciphertextLength).ToArray(), 0);
                }
            }



            SeqTx++;
            BytesTransmitted += (ulong)ciphertextLength;
            return true;
        }
        catch (Exception ex)
        {
#if DEBUG
            ServerMonitor.Log($"[CRYPTO] Encrypt failed: {ex.Message}");
#endif
            ciphertextLength = 0;
            return false;
        }
    }

    public bool EncryptPayloadWithCompression(ReadOnlySpan<byte> plaintext, ReadOnlySpan<byte> aad, Span<byte> result, out int resultLength, out bool wasCompressed)
    {
        try
        {
            wasCompressed = false;
            resultLength = 0;

            Span<byte> encrypted = stackalloc byte[plaintext.Length + 16];
            if (!EncryptPayload(plaintext, aad, encrypted, out int encryptedLength))
                return false;

            if (encryptedLength > 512)
            {
                unsafe
                {
                    fixed (byte* encryptedPtr = encrypted)
                    fixed (byte* resultPtr = result)
                    {
                        int compressedLength = LZ4.Compress(encryptedPtr, encryptedLength, resultPtr, result.Length);
                        if (compressedLength > 0 && compressedLength < encryptedLength)
                        {
                            wasCompressed = true;
                            resultLength = compressedLength;
                            return true;
                        }
                    }
                }
            }

            if (encryptedLength <= result.Length)
            {
                encrypted.Slice(0, encryptedLength).CopyTo(result);
                resultLength = encryptedLength;
                return true;
            }

            return false;
        }
        catch
        {
            wasCompressed = false;
            resultLength = 0;
            return false;
        }
    }

    public bool EncryptPayloadWithSpecificSequence(ReadOnlySpan<byte> plaintext, ReadOnlySpan<byte> aad, ulong sequence, Span<byte> ciphertext, out int ciphertextLength)
    {
        try
        {
            // Use specific sequence for nonce generation
            Span<byte> nonce = stackalloc byte[12];
            BinaryPrimitives.WriteUInt32LittleEndian(nonce, ConnectionId);
            BinaryPrimitives.WriteUInt64LittleEndian(nonce.Slice(4), sequence);

            var cipher = new ChaCha20Poly1305();

            unsafe
            {
                fixed (byte* txKeyPtr = TxKey)
                {
                    var keyParam = new KeyParameter(new ReadOnlySpan<byte>(txKeyPtr, 32).ToArray());
                    var parameters = new AeadParameters(keyParam, 128, nonce.ToArray(), aad.ToArray());

                    cipher.Init(true, parameters);
                    ciphertextLength = cipher.ProcessBytes(plaintext.ToArray(), 0, plaintext.Length, ciphertext.ToArray(), 0);
                    ciphertextLength += cipher.DoFinal(ciphertext.Slice(ciphertextLength).ToArray(), 0);
                }
            }

            // Increment SeqTx to keep crypto sequence in sync
            SeqTx++;
            BytesTransmitted += (ulong)ciphertextLength;
            return true;
        }
        catch (Exception ex)
        {
#if DEBUG
            ServerMonitor.Log($"[CRYPTO] Encrypt failed: {ex.Message}");
#endif
            ciphertextLength = 0;
            return false;
        }
    }

    public bool EncryptPayloadWithSequence(ReadOnlySpan<byte> plaintext, ReadOnlySpan<byte> aad, ulong sequence, Span<byte> result, out int resultLength, out bool wasCompressed)
    {
        try
        {
            wasCompressed = false;
            resultLength = 0;

            // Use specific sequence for nonce generation
            Span<byte> nonce = stackalloc byte[12];
            BinaryPrimitives.WriteUInt32LittleEndian(nonce, ConnectionId);
            BinaryPrimitives.WriteUInt64LittleEndian(nonce.Slice(4), sequence);

            var cipher = new ChaCha20Poly1305();
            Span<byte> encrypted = stackalloc byte[plaintext.Length + 16];

            unsafe
            {
                fixed (byte* txKeyPtr = TxKey)
                {
                    var keyParam = new KeyParameter(new ReadOnlySpan<byte>(txKeyPtr, 32).ToArray());
                    var parameters = new AeadParameters(keyParam, 128, nonce.ToArray(), aad.ToArray());

                    cipher.Init(true, parameters);
                    int encryptedLength = cipher.ProcessBytes(plaintext.ToArray(), 0, plaintext.Length, encrypted.ToArray(), 0);
                    encryptedLength += cipher.DoFinal(encrypted.Slice(encryptedLength).ToArray(), 0);

                    // Increment SeqTx to keep crypto sequence in sync
                    SeqTx++;

                    if (encryptedLength > 512)
                    {
                        unsafe
                        {
                            fixed (byte* encryptedPtr = encrypted)
                            fixed (byte* resultPtr = result)
                            {
                                int compressedLength = LZ4.Compress(encryptedPtr, encryptedLength, resultPtr, result.Length);
                                if (compressedLength > 0 && compressedLength < encryptedLength)
                                {
                                    wasCompressed = true;
                                    resultLength = compressedLength;
                                    return true;
                                }
                            }
                        }
                    }

                    if (encryptedLength <= result.Length)
                    {
                        encrypted.Slice(0, encryptedLength).CopyTo(result);
                        resultLength = encryptedLength;
                        return true;
                    }
                }
            }

            return false;
        }
        catch
        {
            wasCompressed = false;
            resultLength = 0;
            return false;
        }
    }

    public bool DecryptPayload(ReadOnlySpan<byte> ciphertext, ReadOnlySpan<byte> aad, ulong sequence, Span<byte> plaintext, out int plaintextLength)
    {
        try
        {
                                    // Only log for monitoring, not every packet
            if (sequence % 100 == 0) // Log every 100th packet for monitoring
                FileLogger.Log($"[DECRYPT] seq={sequence}, SeqRx={SeqRx}, HighestSeq={_highestSeqReceived}");

            // Log crypto data for comparison (only first packet)
            if (SeqRx == 0 && sequence == 0)
            {
                FileLogger.Log($"=== DECRYPTING FIRST PACKET ===");
                FileLogger.Log($"[SERVER] ConnectionId: {ConnectionId}");
                FileLogger.Log($"[SERVER] RxKey: {GetRxKeyHex()}");
            }

            if (!IsSequenceValid(sequence))
            {
#if DEBUG
                ServerMonitor.Log($"[CRYPTO] Decrypt - Replay attack detected! Sequence: {sequence}");
#endif
                plaintextLength = 0;
                return false;
            }

            Span<byte> nonce = stackalloc byte[12];
            BinaryPrimitives.WriteUInt32LittleEndian(nonce, ConnectionId);
            BinaryPrimitives.WriteUInt64LittleEndian(nonce.Slice(4), sequence);

                        // Log crypto data for comparison (only first packet)
            if (sequence == 0)
            {
                FileLogger.LogHex("[SERVER] Nonce", nonce);
                FileLogger.LogHex("[SERVER] AAD", aad);
                FileLogger.LogHex("[SERVER] Ciphertext", ciphertext);
                FileLogger.Log($"[SERVER] Ciphertext Length: {ciphertext.Length} bytes");
            }

            var cipher = new ChaCha20Poly1305();

            unsafe
            {
                fixed (byte* rxKeyPtr = RxKey)
                {
                    var keyParam = new KeyParameter(new ReadOnlySpan<byte>(rxKeyPtr, 32).ToArray());
                    var parameters = new AeadParameters(keyParam, 128, nonce.ToArray(), aad.ToArray());

                    cipher.Init(false, parameters);

                    // Create arrays for crypto operations
                    byte[] ciphertextArray = ciphertext.ToArray();
                    byte[] plaintextArray = new byte[plaintext.Length];

                    // Log detailed crypto operation for first packet
                    if (sequence == 0)
                    {
                        FileLogger.Log($"[SERVER] Starting decryption...");
                        FileLogger.Log($"[SERVER] Input ciphertext array length: {ciphertextArray.Length}");
                        FileLogger.Log($"[SERVER] Output plaintext array length: {plaintextArray.Length}");
                    }

                    plaintextLength = cipher.ProcessBytes(ciphertextArray, 0, ciphertextArray.Length, plaintextArray, 0);
                    int finalLength = cipher.DoFinal(plaintextArray, plaintextLength);
                    plaintextLength += finalLength;

                    // Log the results for first packet
                    if (sequence == 0)
                    {
                        FileLogger.Log($"[SERVER] ProcessBytes returned: {plaintextLength - finalLength} bytes");
                        FileLogger.Log($"[SERVER] DoFinal returned: {finalLength} bytes");
                        FileLogger.Log($"[SERVER] Total plaintext length: {plaintextLength} bytes");
                        FileLogger.LogHex("[SERVER] Raw decrypted data", new ReadOnlySpan<byte>(plaintextArray, 0, plaintextLength));
                    }

                    // Copy result back to span
                    new ReadOnlySpan<byte>(plaintextArray, 0, plaintextLength).CopyTo(plaintext);
                }
            }

            UpdateReplayWindow(sequence);

            // Log success for first packet
            if (sequence == 0)
            {
                FileLogger.Log($"[SERVER] ✅ DECRYPTION SUCCESS! Plaintext length: {plaintextLength} bytes");
            }

            return true;
        }
        catch (Exception ex)
        {
#if DEBUG
            ServerMonitor.Log($"[CRYPTO] Decrypt failed: {ex.Message}");
#endif
            // Log failure details
            if (sequence == 0)
            {
                FileLogger.Log($"[SERVER] ❌ DECRYPTION FAILED: {ex.Message}");
                FileLogger.Log($"[SERVER] Exception Type: {ex.GetType().Name}");
            }

            plaintextLength = 0;
            return false;
        }
    }

    public bool DecryptPayloadWithDecompression(ReadOnlySpan<byte> data, ReadOnlySpan<byte> aad, ulong sequence, bool isCompressed, Span<byte> plaintext, out int plaintextLength)
    {
        try
        {
            plaintextLength = 0;

            if (isCompressed)
            {
                Span<byte> decompressed = stackalloc byte[data.Length * 4];

                unsafe
                {
                    fixed (byte* dataPtr = data)
                    fixed (byte* decompressedPtr = decompressed)
                    {
                        int decompressedLength = LZ4.Decompress(dataPtr, data.Length, decompressedPtr, decompressed.Length);
                        if (decompressedLength <= 0)
                            return false;

                        return DecryptPayload(decompressed.Slice(0, decompressedLength), aad, sequence, plaintext, out plaintextLength);
                    }
                }
            }
            else
            {
                return DecryptPayload(data, aad, sequence, plaintext, out plaintextLength);
            }
        }
        catch
        {
            plaintextLength = 0;
            return false;
        }
    }

    private bool IsSequenceValid(ulong sequence)
    {
        if (sequence > _highestSeqReceived)
            return true;

        if (sequence + ReplayWindowSize <= _highestSeqReceived)
            return false;

        ulong diff = _highestSeqReceived - sequence;
        if (diff < ReplayWindowSize)
        {
            ulong mask = 1UL << (int)diff;
            return (_replayWindow & mask) == 0;
        }

        return false;
    }

    private void UpdateReplayWindow(ulong sequence)
    {
        if (sequence > _highestSeqReceived)
        {
            ulong shift = sequence - _highestSeqReceived;
            if (shift >= ReplayWindowSize)
            {
                _replayWindow = 1;
            }
            else
            {
                _replayWindow = (_replayWindow << (int)shift) | 1;
            }
            _highestSeqReceived = sequence;
        }
        else
        {
            ulong diff = _highestSeqReceived - sequence;
            if (diff < ReplayWindowSize)
            {
                ulong mask = 1UL << (int)diff;
                _replayWindow |= mask;
            }
        }
    }

    public bool ShouldRekey()
    {
        return BytesTransmitted >= RekeyBytesThreshold ||
               (DateTime.UtcNow - SessionStartTime) >= RekeyTimeThreshold;
    }

    public bool PerformRekey()
    {
        try
        {
            var rekeyContext = new List<byte>();
            rekeyContext.AddRange(Encoding.ASCII.GetBytes("rekey"));
            rekeyContext.AddRange(BitConverter.GetBytes(Math.Max(SeqTx, SeqRx)));

            Span<byte> okm = stackalloc byte[64];
            var hkdf = new HkdfBytesGenerator(new Sha256Digest());

            unsafe
            {
                fixed (byte* saltPtr = SessionSalt)
                {
                    hkdf.Init(new HkdfParameters(new ReadOnlySpan<byte>(saltPtr, 16).ToArray(), null, rekeyContext.ToArray()));
                    hkdf.GenerateBytes(okm);

                    fixed (byte* txKeyPtr = TxKey)
                    fixed (byte* rxKeyPtr = RxKey)
                    {
                        okm.Slice(0, 32).CopyTo(new Span<byte>(txKeyPtr, 32));
                        okm.Slice(32, 32).CopyTo(new Span<byte>(rxKeyPtr, 32));
                    }
                }
            }

            BytesTransmitted = 0;
            SessionStartTime = DateTime.UtcNow;

            SeqTx = 0;
            SeqRx = 0;
            _replayWindow = 0;
            _highestSeqReceived = 0;

            return true;
        }
        catch
        {
            return false;
        }
    }
}

